libs/mutantschemata/cpp_source/rewrite.hpp

```diff
@@ -24,23 +24,171 @@
+
+
+// global clang vars \\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\|
 clang::Rewriter rewriter;
+clang::SourceManager *SourceManager;
+///////////////////////////////////////////////////////////////////////////////|
+
+
+// variables need to uniquely insert meta-mutants \\\\\\\\\\\\\\\\\\\\\\\\\\\\\|
 std::vector<std::string> SourcePaths;
 std::set<std::string> VisitedSourcePaths;
 int mutant_count = 1;
 
-// config mutant schemata (not really implemented)
-bool negateBinaryOperators = true;
-bool negateOverloadedOperators = false;
-bool negateUnaryOperators = false;
+struct InsertedMutant {
+    const clang::FileEntry *fE; // needed to calc FID
+    unsigned exprOffs;
+    unsigned bracketsOffs;
+    
+    InsertedMutant(const clang::FileEntry *f, unsigned int s, unsigned int e): fE(f), exprOffs(s), bracketsOffs(e) {}
+};
+struct MutantInsert {
+    InsertedMutant* mutantLoc;
+    std::string expr;
+    std::string brackets;
+    
+    MutantInsert(InsertedMutant* m, std::string e, std::string b): mutantLoc(m), expr(e), brackets(b) {}
+};
+
+std::vector<MutantInsert*> mutantInserts;
+// custom comperator for mutants based on their FileID (FileEntry) and location (offsets)
+struct mutantLocComp {
+    bool operator() (const MutantInsert *lhs, const MutantInsert *rhs) const {
+        if (lhs->mutantLoc->fE == rhs->mutantLoc->fE) {
+            if (lhs->mutantLoc->exprOffs == rhs->mutantLoc->exprOffs) {
+                return lhs->mutantLoc->bracketsOffs < rhs->mutantLoc->bracketsOffs;
+            }
+            return lhs->mutantLoc->exprOffs < rhs->mutantLoc->exprOffs;
+        }
+        return lhs->mutantLoc->fE < rhs->mutantLoc->fE;
+    }
+};
+// keep an ordered set of the created mutants to prevent duplicates
+// TODO optimise: 1 per file, only overwrite when new mutants were added
+std::set<MutantInsert*, mutantLocComp> insertedMutants;
+///////////////////////////////////////////////////////////////////////////////|
+
+
+// needed functionality not (publically) available in clang \\\\\\\\\\\\\\\\\\\|
+/**
+ * Return true if this character is non-new-line whitespace:
+ * ' ', '\\t', '\\f', '\\v', '\\r'.
+ */
+static inline bool isWhitespaceExceptNL(unsigned char c) {
+    switch (c) {
+        case ' ':
+        case '\t':
+        case '\f':
+        case '\v':
+        case '\r':
+            return true;
+        default:
+            return false;
+    }
+}
+
+class MutantRewriter: public clang::Rewriter {
+public:
+    bool InsertText(const clang::FileEntry *fE, unsigned StartOffs, StringRef Str, bool InsertAfter = true, bool indentNewLines = false) {
+        clang::FileID FID = this->getSourceMgr().translateFile(fE);
+        if (!FID.getHashValue()) {
+            FID = this->getSourceMgr().createFileID(fE, clang::SourceLocation(), clang::SrcMgr::CharacteristicKind::C_User);
+        }
+
+        llvm::SmallString<128> indentedStr;
+        if (indentNewLines && Str.find('\n') != StringRef::npos) {
```

Comment:
suggestion: Even if it looks nice to generate correctly indented code it may complicate the source code.
I would recommend reducing the "formatting" to the bare minimum. If the user need it to look pretty they can use e.g. clang-format.
(I am assuming that indentNewLines mean that the schemata do the indentation)